static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,171 @@
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. We decided that this is the most important software role as too much jitter in timing will cause artifacts and wobbling in the piture.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+{% include figure.html image="/assets/images/fifo-chip-highlight.jpg" width=450 position="right" caption="IDT 7205 FIFO chip" %}
+
+This worked well, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by `CLK_VIS` and the SPI clock signals respectively
+* The SPI only supported clock speeds in power-of-two divisions of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels if the CPU didn't pause writing periodically.
+* If we emitted pixels too slowly, and VGA hardware consumed pixels faster than was being written, then we'd have to start loading pixels before VGA started consuming them.
+* The size of the FIFO effects how much wiggle room we get in the above two cases, especially as the FIFO is too expensive to have enough room for the whole frame.
+* VGA hardware does not read from the FIFO at the pixel clock rate across the entire duration of the frame. Since the visible portion of the frame is only about 73% of the total time spent per frame (because of the blanking area), the average read rate per frame is actually less than the pixel clock rate
+
+The details in how we eneded up on our specific design will be elaberated on in another post, but suffice to say we had to coordinate design details across software, hardware, specific part choices, and part pricing.
+
+### Displaying Pixels
+To display pixels, the `CLK_VIS` signal generated by the VGA timing hardware reads bytes from the FIFO and into a [Programmable Logic Device (PLD)](https://en.wikipedia.org/wiki/Programmable_logic_device) which maps the 8 bits into 9 bits, 3 bits for each of the color channels - red, green, and blue. These 9 bits then get sent to the DAC, described in the following section.
+
+An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. Since there is processing of the read bytes it's possible that the pixels become late compared to `H_SYNC` and `V_SYNC` and this will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.
```

Comment:
```suggestion
An important consideration for this stage is how long it takes to shuttle a byte from the FIFO into the DAC and finally into the monitor. The FIFO and color chip both have propagation delays that might make the read bytes show up "late," and contamination delays that might make them show up "early" compared to `H_SYNC` and `V_SYNC`. This will show up as a horizontal shift. I.E. the "first" pixel could actually show up on the 2nd or 3rd pixel position on the screen, instead of the actual 1st pixel.
```